feat(trace-view): Include errors in the light trace

wmak opened the pull request:
- Refactored the *_map construction to share it between
  transactions/errors
- Accept errors for the event_id now, but the error will still be under
  the `errors` attribute of its corresponding transaction
- Also for the light view getting errors for the current transaction
- Some fixes:
    - Wasn't correctly returning generation for the light view
    - Removing isNull on !event.type:transaction

wmak committed:
feat(trace-view): Include errors in the light trace

- Refactored the *_map construction to share it between
  transactions/errors
- Accept errors for the event_id now, but the error will still be under
  the `errors` attribute of its corresponding transaction
- Also for the light view getting errors for the current transaction
- Some fixes:
    - Wasn't correctly returning generation for the light view
    - Removing isNull on !event.type:transaction

wmak commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -15,14 +16,22 @@
 
 def is_root(item):
-    return item["root"] == "1"
+    return item.get("root", "0") == "1"
```
Did this so we could re-use `construct_span_map` for both errors & transactions, and errors will never be the "root" event

wmak committed:
ref: Removing leftover event.type

Zylphrex commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -134,49 +150,100 @@ def get(self, request, organization, trace_id):
+                    snuba_event["project.id"], snuba_event["id"]
+                )
+            # We didn't get lucky, time to talk to nodestore...
+            for transaction in filter(
+                lambda t: t["transaction"] == error_event["transaction"], transactions
+            ):
```
Might be a personal preference, but I don't like `filter` in `python` where it's unnecessary.

```suggestion
            for transaction in transactions:
                if transaction["transaction"] != error_event["transaction"]:
                    continue
```

wmak replied:
Hmm I personally don't have a preference either way TBH, any specific reason to avoid filter? At least with simple conditions like this they're nearly identical in perf & readability to me

Zylphrex replied:
My argument against `filter` is that these simple conditions evolve over time. We have a tendency to not change more than is necessary so this turns into a giant lambda expression which I would argue is a lot less legible compared to simple if statements. But again, this is just my preference.

Zylphrex commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -134,49 +150,100 @@ def get(self, request, organization, trace_id):
+            for transaction in filter(
+                lambda t: t["transaction"] == error_event["transaction"], transactions
+            ):
+                nodestore_event = eventstore.get_event_by_id(
+                    transaction["project.id"], transaction["id"]
+                )
```
Am I right to assume that this warms the cache so when we're traversing the trace afterwards, it doesn't need to get it from nodestore again?

wmak replied:
That is not the intent, but it is a side effect for the full trace. (We already were making a nodestore call before L197 on master)

Zylphrex commented on src/sentry/api/event_search.py:
```diff
@@ -952,7 +952,8 @@ def convert_search_filter_to_snuba_query(search_filter, key=None, params=None):
 
         is_null_condition = None
-        if search_filter.operator == "!=" and not search_filter.key.is_tag:
+        # TODO(wmak): Skip this for all non-nullable keys not just event.type
```
This is the optimization that was mentioned so we can skip reading an extra `null` mask in clickhouse right?

Zylphrex replied:
As a follow-up, we can potentially apply this to all `has` queries on non-nullable fields.

wmak replied:
:+1: https://getsentry.atlassian.net/browse/VIS-751 Going to be the first thing I tackle once we get the trace view to a decent place

Zylphrex commented on tests/snuba/api/endpoints/test_organization_events_trace.py:
```diff
@@ -269,7 +269,7 @@ def test_root_event(self):
             assert child_event_id in events
             event = events[child_event_id]
-            assert event["generation"] is None
+            assert event["generation"] == 1
```
Why was this `None` before?

wmak replied:
Mistake on my part, when we had a direct line to the root on the quick trace we weren't getting child generations

wmak committed:
ref: Being a bit clearer when things are transacitons vs errors vs events

- Some sneaked in fixes too:
- Realized that naming a tag `trace.*` is a bad idea since we have
  `trace.*` tags already
- Wrapping the error query with a handler

Zylphrex commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -134,49 +156,104 @@ def get(self, request, organization, trace_id):
+        """
+        with sentry_sdk.start_span(op="discover", description="getting trace errors"):
+            with self.handle_query_errors():
+                # Search for errors for the current event if its a transaction, otherwise look for the exact error
```
This won't get the errors on the child transactions like the full trace right? Not a huge deal but if this is true, then the loading states there will have the same problem where it'll appear without errors, and errors will pop in a little later.

wmak replied:
It won't, I'll backlog something to do the best effort association (ie. the error events trace.span == transaction.span) but adding errors is already pretty slow cause of the 2x snuba query and I try to keep the errors query light ([either txn name or id](https://github.com/getsentry/sentry/pull/24590/files#diff-619b86f549a9f11a566ca757be8da43a63bcaf6e4d7d3fe88a4222937a0209c6R214-R217)) so this best effort association isn't possible until I consolidate the snuba queries (also a backlogged task)

Zylphrex reviewed (approved):
Just the one question, otherwise LGTM
